### PR TITLE
Enable navigation from historical file to solution

### PR DESCRIPTION
**Pull request by jcansdale**
### What this PR does

Extend the `GoToSolutionOrPullRequestFile` command to support navigation from a file opened using the history view to the same file in the users solution/working directory.

### The history of a file

1. Right click on a source file from a repository and `Code context > View History...`
image

2. Double click on a revision in the history
image

3. Right click on the read-only historical file and `Open File in Solution`
image

4. It should navigate to the equivalent position in your solution / working directory

### The commit history

1. Select `View History...` from the status bar
image

2. Double click on a commit in the history
image

3. Double click on a changed file in the commit
image

4. Right click on the commit diff view and `Open File in Solution`
image

4. It should navigate to the equivalent position in your solution / working directory

### CODE REVIEW

File: src/GitHub.App/Services/GitHubContextService.cs [hunk 1]
```diff
@@ -305,6 +307,55 @@ public bool TryOpenFile(string repositoryDir, GitHubContext context)
             }
         }
 
+        /// <inheritdoc/>
+        public string FindObjectishForTFSTempFile(string tempFile)
+        {
+            var match = tempFileObjectishRegex.Match(tempFile);
+            if (match.Success)
+            {
```
Comment (StanleyGoldman) [inline src/GitHub.App/Services/GitHubContextService.cs]: Why did you name the group `objectish`?

File: src/GitHub.App/Services/GitHubContextService.cs [hunk 2]
```diff
@@ -305,6 +307,55 @@ public bool TryOpenFile(string repositoryDir, GitHubContext context)
...
+                return match.Groups["objectish"].Value;
+            }
+
```
Comment (StanleyGoldman) [inline src/GitHub.App/Services/GitHubContextService.cs]: Test single comment

File: src/GitHub.App/Services/GitHubContextService.cs (hunk 1 above)
Comment (StanleyGoldman) [inline src/GitHub.App/Services/GitHubContextService.cs]: Why did you name the group `objectish`?

File: src/GitHub.App/Services/GitHubContextService.cs [hunk 3]
```diff
@@ -305,6 +307,55 @@ public bool TryOpenFile(string repositoryDir, GitHubContext context)
             }
         }
 
+        /// <inheritdoc/>
+        public string FindObjectishForTFSTempFile(string tempFile)
+        {
+            var match = tempFileObjectishRegex.Match(tempFile);
+            if (match.Success)
+            {
+                return match.Groups["objectish"].Value;
```
Comment (StanleyGoldman) [inline src/GitHub.App/Services/GitHubContextService.cs]: Why did you name the group `objectish`?

Comment (StanleyGoldman) [reply src/GitHub.App/Services/GitHubContextService.cs]: `(first 8 chars of a blob SHA)` ahh

File: src/GitHub.App/Services/GitHubContextService.cs [hunk 4]
```diff
@@ -305,6 +307,55 @@ public bool TryOpenFile(string repositoryDir, GitHubContext context)
...
+                var blob = repo.Lookup<Blob>(objectish);
+                if (blob == null)
+                {
+                    return (null, null);
+                }
+
+                foreach (var commit in repo.Commits)
+                {
+                    var trees = new Stack<Tree>();
+                    trees.Push(commit.Tree);
```
Comment (StanleyGoldman) [inline src/GitHub.App/Services/GitHubContextService.cs]: I now know what a Git blob sha is, I've yet to wrap my head around a Commit Tree.

File: src/GitHub.App/Services/GitHubContextService.cs (hunk 3 above)
Comment (jcansdale) [reply src/GitHub.App/Services/GitHubContextService.cs]: It started off being called blobish, but I stole the name from the `Lookup` method you found. 😄

File: src/GitHub.VisualStudio/Commands/GoToSolutionOrPullRequestFileCommand.cs [hunk 5]
```diff
@@ -198,6 +216,81 @@ void OnBeforeQueryStatus(object sender, EventArgs e)
...
+            if (path == null)
+            {
+                return false;
+            }
+
+            var objectish = gitHubContextService.Value.FindObjectishForTFSTempFile(path);
+            if (objectish == null)
+            {
+                return false;
+            }
```
Comment (StanleyGoldman) [inline src/GitHub.VisualStudio/Commands/GoToSolutionOrPullRequestFileCommand.cs]: The first three if statements from this function and the function above are meant to be the same.
The only minor minor difference _(took me a moment to spot)_ is you are saving the repository dir in the second implementation.

The intent might be clearer if you combine both to one function that returns `repositoryDir` if the conditions are right or `null` if they fail.

Comment (StanleyGoldman) [reply src/GitHub.VisualStudio/Commands/GoToSolutionOrPullRequestFileCommand.cs]: `GetRepositoryDirectoryForSourceViewAndPath` or something..

File: src/GitHub.App/Services/GitHubContextService.cs (hunk 4 above)
Comment (StanleyGoldman) [reply src/GitHub.App/Services/GitHubContextService.cs]: I have to read the tests more

**Comment by StanleyGoldman** [comment]
Everything works here. Just one thing that can be clarified i spotted.
I still feel like I need to read the tests more to understand what a Commit Tree is.

**Pull request merged**